specifyweb/frontend/js_src/lib/components/QueryBuilder/CheckReadAccess.tsx

```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { commonText } from '../../localization/common';
+import { queryText } from '../../localization/query';
+import { RA } from '../../utils/types';
+import { Ul } from '../Atoms';
+import { SerializedResource } from '../DataModel/helperTypes';
+import { SpQuery, Tables } from '../DataModel/types';
+import { Dialog } from '../Molecules/Dialog';
+import { TableIcon } from '../Molecules/TableIcon';
+import { hasTablePermission } from '../Permissions/helpers';
+import { QueryFieldSpec } from './fieldSpec';
+
+export function CheckReadAccess({
+  query,
+}: {
+  readonly query: SerializedResource<SpQuery>;
+}): JSX.Element | null {
+  const [noAccessTables, setNoAccessTables] = React.useState<RA<keyof Tables>>(
+    []
+  );
+
+  function accessTables(query: SerializedResource<SpQuery>): RA<keyof Tables> {
+    const info = query.fields.flatMap((field) => {
+      const fieldSpec = QueryFieldSpec.fromStringId(
+        field.stringId,
+        field.isRelFld ?? false
+      );
+      const relatedTable = fieldSpec.joinPath.map(({ model }) => model.name);
+      return relatedTable;
+    });
+
+    const withoutDuplicates = new Set(info);
+
+    const noAccessTables = Array.from(withoutDuplicates).filter(
```

Comment:
(optional) instead of `Array.from(new Set(someVar))` can use our util and just type `f.unique(someVar)`